File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int f(vector<vector<int>>& dp , vector<int>& nums,  int i , int buy){
        int n = nums.size();
        if(i >= n) return 0;
        if(dp[i][buy] != -1) return dp[i][buy];
        int take = 0;
        if(buy == 0){
            take = max(-nums[i] + f(dp , nums , i + 1 , 1)  , f(dp , nums, i +1 , 0));
        }else{
            take = max(nums[i] + f(dp , nums , i + 1 , 0) , f(dp , nums, i+ 1 , 1));
        }
        int notTake = f(dp , nums , i + 1 , buy);
        return dp[i][buy] = max(take , notTake);
    }
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<vector<int>> dp(n , vector<int>(2 , -1));
        return f(dp , prices , 0 , 0);
    }
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] nums) {
        int n = nums.length;
        int[] cur = new int[n + 1];
        int[] next = new int[n + 1];
        for(int i = n - 1 ; i >= 0 ; i--){
            for(int j = 0  ; j < 2 ; j++){
                int take = 0;
                if(j == 0){
                    take = Math.max( next[1] - nums[i] , next[0]);
                }else{
                    take = Math.max(nums[i] + next[0] , next[1]);
                }
                int notTake = next[j];
                cur[j] = Math.max(take , notTake);
            }
            next = cur;
        }
        return cur[0];
    }
}